### PR TITLE
feat: add ledger methods for faucet spending

### PR DESCRIPTION
**Pull request by hansieodendaal**
Description
---
- Added ledger methods to support faucet spending.
- Consolidated h/w ledger support into the `minotari_ledger_wallet_comms` crate.
- Added test methods for all h/w ledger functions in the form of an example.
- Addedd a seperate crate for common types shared between the Ledger application and the rest of the code base.

~~**Note:** I do not think the ledger feature (`#[cfg(feature = "ledger")]`) is nesseccary anymore. Any opinions?~~

Motivation and Context
---
Faucet spending was not enabled on ledger devices.

How Has This Been Tested?
---
Compiled the ledger project `minotari_ledger_wallet`, uploaded the binaries to a ledger nano s device, then ran `cargo run --release --example ledger_demo`.

```
Transport created in 103.1686ms
Transport created in 5.8315ms
Transport created in 6.0115ms
Transport created in 5.9472ms
Transport created in 5.9368ms
Transport created in 5.8717ms
Transport created in 5.9897ms
Transport created in 6.0287ms
Transport created in 5.9255ms
Transport created in 5.9359ms

Application verified in 1.293162s
Application verified in 200ns
Application verified in 200ns
Application verified in 100ns
Application verified in 200ns
Application verified in 400ns
Application verified in 10.2µs
Application verified in 200ns
Application verified in 100ns
Application verified in 100ns


test: GetAppName
app name:       minotari_ledger_wallet

test: GetVersion
version:        1.0.0-pre.16

test: GetPublicAlpha
public_alpha:   ea10ad3d125beb2c1ebfa2d295ee8e42fd4a85ad831886f8b0b408f8390a9e3c

test: GetPublicKey
public_key:     4a01ab9c27a256827c6ee85373412f80cc88040425afe6301c9b1c7773356e2a

test: GetScriptSignature
script_sig:     (3058744ade0c70a5fe914ed64695c82bc4c71477f8df46578fb3780c06c8110a,5e3fb09be68a31f397330037b6fc5b09227f1a89f74c66fe72fa34c507b90e2b,ed1a3a74f17ae19d70252a2d0ca92a0295f1f0c43565653b4f1fea9ffac32c0d,4f73a5c7cb723b12eca2c3ff57b8156ccacf9ed2790ac2b2e6c033147536e30d,7e8abeb6774fa9e63b181b54e47f09198116d02ec2c1f536c977cf8254d69a09)

test: GetScriptOffset
script_offset:  3a46a979ec9c2d94fbb55e09221fd0deb9e869dc849975e5ee163fb36bbe0709

test: GetViewKey
view_key:       158e9d986d15c40ac3f6178f5947fde0cbd4f71f7de01c00c88cfa570e648e09

test: GetDHSharedSecret
shared_secret:  e85e660bd4d55f4d57784c3592e1ed063c77b8b1aa37431c078e8a7e18c8d23a

test: GetRawSchnorrSignature
signature:      (2971d25b5920bc7a102a9d22c652fb1cf8da85ef08b8464dca6a363612040903,0a7f70764654239c51590d0f32b3b23368e355bb269d99cc151e1c261db56156)

test: GetScriptSchnorrSignature
signature:      (c1fce9daf50cc65e69440f3f4383db1f64b1fd190bdcf7824c9dce0dcd68c208,ace0f952697b2496e60deadee0d40a23c263fa0d73c66a2742767fadad61fe33)

test: Ledger app not running
✔ Exit the 'MinoTari Wallet' Ledger app and press Enter to continue.. · Ok

test: Ledger disconnected
✔ Disconnect the Ledger device and press Enter to continue.. · Ok

test: Ledger reconnected
✔ Reconnect the Ledger device (with password) and press Enter to continue.. · Ok

test: Ledger app restart
✔ Start the 'MinoTari Wallet' Ledger app and press Enter to continue.. · Ok
view_key:       158e9d986d15c40ac3f6178f5947fde0cbd4f71f7de01c00c88cfa570e648e09

Test completed successfully
```

What process can a PR reviewer use to test or verify this change?
---
- Code review.
- Compile the ledger project `minotari_ledger_wallet`, upload the binaries to a ledger nano s device, then run `cargo run --release --example ledger_demo`.

Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

### CODE REVIEW

File: applications/minotari_console_wallet/src/automation/commands.rs [hunk 1]
```diff
@@ -1007,7 +1007,7 @@ pub async fn command_runner(
 
                 let mut metadata_signature = Signature::default();
                 match key_manager_service
-                    .sign_with_nonce_and_message(
+                    .sign_with_challenge_and_message(
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/automation/commands.rs]:
```suggestion
                    .sign_with_nonce_and_challenge(
```
Previous name was incorrect, but this is more incorrect. 

in `sig = r +e*k`
you have 
r = nonce
e = challenge

where challenge = hash(R||K||message)

File: applications/minotari_console_wallet/src/automation/commands.rs [hunk 2]
```diff
@@ -971,7 +971,7 @@ pub async fn command_runner(
 
                 let mut script_signature = Signature::default();
                 match key_manager_service
-                    .sign_with_nonce_and_message(
+                    .sign_with_challenge_and_message(
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/automation/commands.rs]:
```suggestion
                    .sign_with_nonce_and_challenge(
```

File: applications/minotari_ledger_wallet/comms/src/ledger_wallet.rs [hunk 3]
```diff
@@ -39,12 +40,65 @@ pub enum Instruction {
...
 
+#[repr(u16)]
+#[derive(FromPrimitive, Debug, Copy, Clone, PartialEq)]
+pub enum AppSW {
+    Deny = 0xB001,
+    WrongP1P2 = 0xB002,
+    InsNotSupported = 0xB003,
+    ClaNotSupported = 0xB004,
+    ScriptSignatureFail = 0xB005,
+    MetadataSignatureFail = 0xB006,
```
Comment (SWvheerden) [inline applications/minotari_ledger_wallet/comms/src/ledger_wallet.rs]: is this called?
I see the instruction above removed

File: applications/minotari_ledger_wallet/wallet/src/handlers/get_schnorr_signature.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    let mut nonce_bytes = [0u8; 32];
+    nonce_bytes.clone_from_slice(&data[24..56]);
+
+    let random_nonce_a = get_random_nonce()?.deref().clone();
+    let random_nonce_b = get_random_nonce()?.deref().clone();
+    if random_nonce_a == random_nonce_b {
+        SingleMessage::new("Nonces not unique!").show_and_wait();
+        return Err(AppSW::SchnorrSignatureFail);
+    }
```
Comment (SWvheerden) [inline applications/minotari_ledger_wallet/wallet/src/handlers/get_schnorr_signature.rs]:
```suggestion
```
The ledger is slow and this ain't cheap. 
The check that it does not return default should be sufficient here. 
Its not required just try a second nonce and see if its the same as before.

File: applications/minotari_ledger_wallet/comms/src/ledger_wallet.rs [hunk 5]
```diff
@@ -39,12 +40,65 @@ pub enum Instruction {
...
-    GetDHSharedSecret = 0x10,
+    GetViewKey = 0x07,
+    GetDHSharedSecret = 0x08,
+    GetRawSchnorrSignature = 0x09,
+    GetScriptSchnorrSignature = 0x10,
 }
 
+#[repr(u16)]
+#[derive(FromPrimitive, Debug, Copy, Clone, PartialEq)]
+pub enum AppSW {
```
Comment (SWvheerden) [inline applications/minotari_ledger_wallet/comms/src/ledger_wallet.rs]: this is delcared twice?

File: applications/minotari_ledger_wallet/wallet/src/main.rs [hunk 6]
```diff
@@ -87,21 +89,24 @@ unsafe impl critical_section::Impl for MyCriticalSection {
 // Application status words.
 #[repr(u16)]
 pub enum AppSW {
```
Comment (SWvheerden) [inline applications/minotari_ledger_wallet/wallet/src/main.rs:96]: declared twice?

File: applications/minotari_ledger_wallet/comms/src/accessor_methods.rs [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+                Err(e) => return Err(LedgerDeviceError::Processing(format!("1 GetVersion: {}", e))),
+            };
+            Ok(name.to_string())
+        },
+        Err(e) => Err(LedgerDeviceError::Processing(format!("2 GetVersion: {}", e))),
+    }
+}
+
+/// Get the public alpha key from the ledger device
+pub fn ledger_get_public_alpha(account: u64) -> Result<PublicKey, LedgerDeviceError> {
```
Comment (SWvheerden) [inline applications/minotari_ledger_wallet/comms/src/accessor_methods.rs]: this instruction should be spend, not alpha

File: applications/minotari_ledger_wallet/wallet/src/main.rs [hunk 8]
```diff
@@ -144,9 +152,13 @@ impl KeyType {
     fn from_branch_key(n: u64) -> Result<Self, AppSW> {
         // These numbers need to match the TransactionKeyManagerBranches in:
         // base_layer/core/src/transactions/key_manager/interface.rs
+        // See `pub enum TransactionKeyManagerBranch` in
+        // `tari_wallet/src/transaction_service/transaction_key_manager.rs` for the mapping of the branch key to
+        // the key type.
         match n {
-            7 => Ok(Self::Spend),
             6 => Ok(Self::OneSidedSenderOffset),
+            7 => Ok(Self::Spend),
```
Comment (SWvheerden) [inline applications/minotari_ledger_wallet/wallet/src/main.rs]: Is this correct?
I dont think spend should be here?

File: base_layer/common_types/src/lib.rs [hunk 9]
```diff
@@ -30,6 +30,7 @@ pub mod emoji;
 pub mod encryption;
 pub mod epoch;
 pub mod grpc_authentication;
+pub mod key_manager;
```
Comment (SWvheerden) [inline base_layer/common_types/src/lib.rs]:
```suggestion
pub mod key_branches;
```
Lets be specific what this is

File: base_layer/core/src/transactions/key_manager/interface.rs [hunk 10]
```diff
@@ -279,7 +226,7 @@ pub trait TransactionKeyManagerInterface: KeyManagerInterface<PublicKey> {
         challenge: &[u8],
     ) -> Result<CheckSigSchnorrSignature, TransactionError>;
 
-    async fn sign_with_nonce_and_message(
+    async fn sign_with_challenge_and_message(
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/interface.rs]:
```suggestion
    async fn sign_with_nonce_and_challenge(
```

File: base_layer/core/src/transactions/key_manager/wrapper.rs [hunk 11]
```diff
@@ -482,7 +482,7 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
         self.transaction_key_manager_inner
             .read()
             .await
-            .sign_with_nonce_and_message(private_key_id, nonce, challenge)
+            .sign_with_challenge_and_message(private_key_id, nonce, challenge)
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/wrapper.rs]:
```suggestion
            .sign_with_nonce_and_challenge(private_key_id, nonce, challenge)
```

File: base_layer/core/src/transactions/transaction_components/wallet_output.rs [hunk 12]
```diff
@@ -276,7 +276,7 @@ impl WalletOutput {
             &message,
         );
         let script_key_partial_script_signature = key_manager
-            .sign_with_nonce_and_message(&self.script_key_id, &ephemeral_public_key_self.key_id, &challenge)
+            .sign_with_challenge_and_message(&self.script_key_id, &ephemeral_public_key_self.key_id, &challenge)
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/wallet_output.rs]: same as others

File: base_layer/core/src/transactions/transaction_components/wallet_output_builder.rs [hunk 13]
```diff
@@ -256,7 +259,7 @@ impl WalletOutputBuilder {
             &metadata_message,
         );
         let sender_partial_metadata_signature_self = key_manager
-            .sign_with_nonce_and_message(sender_offset_key_id, &ephemeral_pubkey_self.key_id, &challenge)
+            .sign_with_challenge_and_message(sender_offset_key_id, &ephemeral_pubkey_self.key_id, &challenge)
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/wallet_output_builder.rs]: same as others

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 14]
```diff
@@ -25,15 +25,20 @@ use blake2::Blake2b;
...
+    ledger_get_dh_shared_secret,
+    ledger_get_public_key,
+    ledger_get_raw_schnorr_signature,
+    ledger_get_script_offset,
+    ledger_get_script_schnorr_signature,
+    ledger_get_script_signature,
 };
 use rand::rngs::OsRng;
-use strum::IntoEnumIterator;
 #[cfg(feature = "ledger")]
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs:37]: is this correct still?

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 15]
```diff
@@ -182,87 +190,118 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                branch, self.wallet_type
+            ))),
             Some(_) => Ok(KeyId::Managed {
                 branch: branch.to_string(),
                 index: 0,
             }),
         }
     }
 
+    #[allow(clippy::too_many_lines)]
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]: this correct?
The function seems smaller?

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 16]
```diff
@@ -182,87 +190,118 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                                .clone()
+                                .ok_or(KeyManagerServiceError::LedgerViewKeyInaccessible)?;
+                            Ok(PublicKey::from_secret_key(&view_key))
+                        },
+                        _ => Err(KeyManagerServiceError::BranchNotSupported(format!(
+                            "{}, {}",
+                            branch, self.wallet_type
+                        ))),
+                    },
+                    _ => {
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs:272]: this is not correct, we need to handle the provided wallet as well. But the old code was broken as well, so this is fine as is, we can fix it in a new pr. This pr is already long enough

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 17]
```diff
@@ -332,14 +375,20 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                            branch, self.wallet_type
+                        )))?
                         .read()
                         .await;
                     let private_alpha = km.get_private_key(0)?;
                     let km = self
                         .key_managers
                         .get(branch)
-                        .ok_or(KeyManagerServiceError::UnknownKeyBranch)?
+                        .ok_or(KeyManagerServiceError::UnknownKeyBranch(format!(
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]: perhaps we need a self helper function to help here.

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 18]
```diff
@@ -786,10 +854,10 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
         let commitment = self.get_commitment(commitment_mask_key_id, value).await?;
         let commitment_private_key = self.get_private_key(commitment_mask_key_id).await?;
 
-        #[allow(unused_variables)] // When ledger isn't enabled
         match (&self.wallet_type, script_key_id) {
             (
                 WalletType::Ledger(ledger),
+                // TODO: Is this correct? Should it not be 'KeyId::Managed'?
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]:
```suggestion
```
yes
And we should not have todos in code

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 19]
```diff
@@ -1154,25 +1164,104 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
     pub async fn sign_script_message(
         &self,
         private_key_id: &TariKeyId,
-        challenge: &[u8],
+        nonce: &[u8],
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]:
```suggestion
        challange: &[u8],
```
This is wrong, challenge is correct

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 20]
```diff
@@ -1154,25 +1164,104 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                let private_key = self.get_private_key(private_key_id).await?;
+                let signature = CheckSigSchnorrSignature::sign(&private_key, nonce, &mut OsRng)?;
+
+                Ok(signature)
+            },
+        }
     }
 
-    pub async fn sign_with_nonce_and_message(
+    pub async fn sign_with_challenge_and_message(
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]: same as the others

**Review by SWvheerden — COMMENTED** [review]
This looks good
A few missed name changes but the key concept here is correct

File: applications/minotari_ledger_wallet/wallet/src/main.rs (hunk 8 above)
Comment (hansieodendaal) [reply applications/minotari_ledger_wallet/wallet/src/main.rs]: It comes from PR #6413

File: base_layer/core/src/transactions/key_manager/inner.rs (hunk 14 above)
Comment (hansieodendaal) [reply base_layer/core/src/transactions/key_manager/inner.rs:37]: Yes, still used:
```
error[E0599]: no variant or associated item named `iter` found for enum `TransactionKeyManagerBranch` in the current scope                                                                                                                      
   --> base_layer\core\src\transactions\key_manager\inner.rs:136:52
    |
136 |         for branch in TransactionKeyManagerBranch::iter() {
    |                                                    ^^^^ variant or associated item not found in `TransactionKeyManagerBranch`
    |
note: if you're trying to build a new `TransactionKeyManagerBranch`, consider using `TransactionKeyManagerBranch::from_key` which returns `TransactionKeyManagerBranch`
   --> C:\Users\pluto\.tari\Code\@tari-project\base_layer\common_types\src\key_branches.rs:69:5
    |
69  |     pub fn from_key(key: &str) -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the trait is in scope
help: trait `IntoEnumIterator` which provides `iter` is implemented but not in scope; perhaps you want to import it
    |
22  + use strum::IntoEnumIterator;
    |
```

File: applications/minotari_ledger_wallet/wallet/src/main.rs (hunk 6 above)
Comment (hansieodendaal) [reply applications/minotari_ledger_wallet/wallet/src/main.rs:96]: Yes, not ideal. To solve this I created a seperate crate for common types shared between the Ledger application and the rest of the code base.

File: applications/minotari_ledger_wallet/comms/src/ledger_wallet.rs (hunk 5 above)
Comment (hansieodendaal) [reply applications/minotari_ledger_wallet/comms/src/ledger_wallet.rs]: Yes, not ideal. To solve this I created a seperate crate for common types shared between the Ledger application and the rest of the code base.

**Review by SWvheerden — APPROVED** [review]

**Pull request merged**